### PR TITLE
[expo-cli] changed error handling for ios upload to use fastlane exit code

### LINKED ISSUE #2368: expo upload:ios on github actions (using macos-latest) succeeds and fails!
## Description

`expo upload:ios` on github actions (using `macos-latest`) succeeds and fails!

## Expected Behavior

**I ran:** 
```
	expo upload:ios --latest --non-interactive \
		--apple-id $(APPLE_ID) \
		--apple-id-password $(APPLE_ID_PASS)
```
**I expected** the latest built app to be uploaded to itunes connect and the process to complete successfully.

## Observed Behavior

* command fails:
```
2020-07-16T14:50:35.7993480Z [14:50:35]: [32m--------------------------------------------------------------------[0m
2020-07-16T14:50:35.7994550Z [14:50:35]: [32mSuccessfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.[0m
2020-07-16T14:50:35.7995410Z [14:50:35]: [32m--------------------------------------------------------------------[0m
2020-07-16T14:50:35.8158710Z [14:50:35]: [32mSuccessfully uploaded the new binary to App Store Connect[0m
2020-07-16T14:50:35.8160500Z [14:50:35]: [33m`skip_waiting_for_build_processing` used and no `changelog` supplied - skipping waiting for build processing[0m
2020-07-16T14:50:35.8637030Z [14:50:35] Failed to upload the standalone app to the app store.
2020-07-16T14:50:35.8639080Z [14:50:35] Unexpected token N in JSON at position 0
2020-07-16T14:50:35.8648380Z [14:50:35] SyntaxError: Unexpected token N in JSON at position 0
2020-07-16T14:50:35.8648780Z     at JSON.parse (<anonymous>)
2020-07-16T14:50:35.8650070Z     at runFastlaneAsync (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload/utils.ts:76:20)
2020-07-16T14:50:35.8650390Z     at processTicksAndRejections (internal/process/task_queues.js:97:5)
2020-07-16T14:50:35.8651280Z     at IOSUploader._uploadToTheStore (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload/IOSUploader.ts:225:5)
2020-07-16T14:50:35.8652530Z     at IOSUploader.upload (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload/BaseUploader.ts:33:5)
2020-07-16T14:50:35.8653440Z     at /Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload.ts:119:9
2020-07-16T14:50:35.8654480Z     at Command.<anonymous> (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/exp.ts:80:7)
```
* app is uploaded to app store (i consistently get a notification from testflight roughly 20mins after the build fails)

Full dump of logs:
```
[36;1mmake upload-ios[0m
2020-07-16T14:35:29.4534650Z shell: /bin/bash -e {0}
2020-07-16T14:35:29.4534910Z env:
2020-07-16T14:35:29.4535130Z   ENVIRONMENT: prod
2020-07-16T14:35:29.4535910Z   APP_CONFIG: ***

2020-07-16T14:35:29.4536870Z   FASTLANE_SESSION: ***
2020-07-16T14:35:29.4537550Z   FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD: ***
2020-07-16T14:35:29.4537990Z   APPLE_ID: ***
2020-07-16T14:35:29.4538320Z   APPLE_ID_PASS: ***
2020-07-16T14:35:29.4538610Z ##[endgroup]
2020-07-16T14:35:29.5113390Z expo upload:ios --latest --non-interactive \
2020-07-16T14:35:29.5120400Z 		--apple-id *** \
2020-07-16T14:35:29.5121510Z 		--apple-id-password ***
2020-07-16T14:35:31.0267830Z [14:35:31] Downloading build from https://exp-shell-app-assets.s3.us-west-1.amazonaws.com/ios/%40***/xxxx-archive.ipa
2020-07-16T14:35:38.2459910Z 
2020-07-16T14:35:38.2559460Z [14:35:38] Resolving the ITC team ID...
2020-07-16T14:35:56.8767760Z [14:35:56] ITC team ID is XXXXX
2020-07-16T14:35:56.8768300Z [14:35:56] Ensuring the app exists on App Store Connect, this may take a while...
2020-07-16T14:36:03.9429820Z 
2020-07-16T14:36:03.9431420Z +----------------+--------------------+
2020-07-16T14:36:03.9432400Z |     [32mSummary for produce 2.146.1[0m     |
2020-07-16T14:36:03.9433090Z +----------------+--------------------+
2020-07-16T14:36:03.9433800Z | app_identifier | xxxxx              |
2020-07-16T14:36:03.9434780Z | app_name       | xxxxx              |
2020-07-16T14:36:03.9436010Z | username       | ***                |
2020-07-16T14:36:03.9436360Z | language       | English            |
2020-07-16T14:36:03.9436680Z | sku            | xxxxx              |
2020-07-16T14:36:03.9436960Z | platform       | ios                |
2020-07-16T14:36:03.9437280Z | skip_itc       | false              |
2020-07-16T14:36:03.9437560Z | skip_devcenter | false              |
2020-07-16T14:36:03.9439100Z +----------------+--------------------+
2020-07-16T14:36:03.9439220Z 
2020-07-16T14:36:03.9440120Z [14:36:03]: [32m[DevCenter] App 'xxxx' already exists, nothing to do on the Dev Center[0m
2020-07-16T14:36:13.2222010Z [14:36:13]: [32mApp 'xxxx' already exists (xxxx), nothing to do on App Store Connect[0m
2020-07-16T14:36:14.9292580Z [14:36:14] Uploading the app to Testflight, hold tight...
2020-07-16T14:36:16.0282780Z [14:36:16]: Login to App Store Connect (***)
2020-07-16T14:36:20.1802760Z [14:36:20]: Login successful
2020-07-16T14:36:22.5836780Z [14:36:22]: [32mReady to upload new build to TestFlight (App: xxxx)...[0m
2020-07-16T14:36:24.8983880Z [14:36:24]: Fetching password for transporter from environment variable named `FASTLANE_APPLE_APPLICATION_SPECIFIC_PASSWORD`
2020-07-16T14:36:24.8985170Z [14:36:24]: Going to upload updated app to App Store Connect
2020-07-16T14:36:24.8988030Z [14:36:24]: [32mThis might take a few minutes. Please don't interrupt the script.[0m
2020-07-16T14:42:17.9454590Z [14:42:17]: [31m[Transporter Error Output]: Failed to upload file: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20200716-5663-1r3u5wr/xxxx.itmsp/xxxx.ipa
2020-07-16T14:42:17.9455550Z [0m
2020-07-16T14:50:35.2899200Z [14:50:35]: [32miTunes Transporter successfully finished its job[0m
2020-07-16T14:50:35.4786400Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main> DBG-X: JSON-START>>{"dev-id-results":{"upload_id":"xxxx"}}<<JSON-END
2020-07-16T14:50:35.4787380Z [0m
2020-07-16T14:50:35.4788640Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main> DBG-X: Unlocking token at path /Users/runner/Library/Caches/com.apple.amp.itmstransporter/UploadTokens/Mac-xxxx_xxxx_xxxx.itmsp.token
2020-07-16T14:50:35.4789330Z [0m
2020-07-16T14:50:35.4821430Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main> DBG-X: Releasing token file lock
2020-07-16T14:50:35.4822390Z [0m
2020-07-16T14:50:35.4823680Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main> DBG-X: Removing token at path /Users/runner/Library/Caches/com.apple.amp.itmstransporter/UploadTokens/Mac-xxxx_xxxx_xxxx.itmsp.token
2020-07-16T14:50:35.4824260Z [0m
2020-07-16T14:50:35.4825630Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main>  INFO: Done performing uploadDone notification to Apple.
2020-07-16T14:50:35.4826260Z [0m
2020-07-16T14:50:35.4828090Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main>  INFO: Transporter's command line arguments are: -m upload -u *** -p **hidden value** -f /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20200716-5663-1r3u5wr/xxxx.itmsp -t DAV,Signiant -k 100000
2020-07-16T14:50:35.4828810Z [0m
2020-07-16T14:50:35.4829940Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main>  INFO: The package: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20200716-5663-1r3u5wr/xxxx.itmsp has been successfully uploaded.
2020-07-16T14:50:35.4830690Z [0m
2020-07-16T14:50:35.4831310Z [14:50:35]: [33m[iTMSTransporter] 
2020-07-16T14:50:35.4831900Z [0m
2020-07-16T14:50:35.4832500Z [14:50:35]: [33m[iTMSTransporter] 
2020-07-16T14:50:35.4833060Z [0m
2020-07-16T14:50:35.4833950Z [14:50:35]: [33m[iTMSTransporter] 
2020-07-16T14:50:35.4834550Z [0m
2020-07-16T14:50:35.4835310Z [14:50:35]: [33m[iTMSTransporter] Package Summary:
2020-07-16T14:50:35.4835900Z [0m
2020-07-16T14:50:35.4836530Z [14:50:35]: [33m[iTMSTransporter]  
2020-07-16T14:50:35.4837100Z [0m
2020-07-16T14:50:35.4837830Z [14:50:35]: [33m[iTMSTransporter] 1 packages were uploaded successfully:
2020-07-16T14:50:35.4838410Z [0m
2020-07-16T14:50:35.4839290Z [14:50:35]: [33m[iTMSTransporter] 	/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20200716-5663-1r3u5wr/xxxx.itmsp
2020-07-16T14:50:35.4839820Z [0m
2020-07-16T14:50:35.4840790Z [14:50:35]: [33m[iTMSTransporter] [2020-07-16 14:50:35 UTC] <main> DBG-X: Returning 0
2020-07-16T14:50:35.4841910Z [0m
2020-07-16T14:50:35.4842260Z [14:50:35]: iTunes Transporter output above ^
2020-07-16T14:50:35.4843350Z [14:50:35]: [31mFailed to upload file: /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/d20200716-5663-1r3u5wr/xxxx.itmsp/xxxx.ipa
2020-07-16T14:50:35.4843970Z [0m
2020-07-16T14:50:35.4844770Z [14:50:35]: [33mAlthough errors occurred during execution of iTMSTransporter, it returned success status.[0m
2020-07-16T14:50:35.7993480Z [14:50:35]: [32m--------------------------------------------------------------------[0m
2020-07-16T14:50:35.7994550Z [14:50:35]: [32mSuccessfully uploaded package to App Store Connect. It might take a few minutes until it's visible online.[0m
2020-07-16T14:50:35.7995410Z [14:50:35]: [32m--------------------------------------------------------------------[0m
2020-07-16T14:50:35.8158710Z [14:50:35]: [32mSuccessfully uploaded the new binary to App Store Connect[0m
2020-07-16T14:50:35.8160500Z [14:50:35]: [33m`skip_waiting_for_build_processing` used and no `changelog` supplied - skipping waiting for build processing[0m
2020-07-16T14:50:35.8637030Z [14:50:35] Failed to upload the standalone app to the app store.
2020-07-16T14:50:35.8639080Z [14:50:35] Unexpected token N in JSON at position 0
2020-07-16T14:50:35.8648380Z [14:50:35] SyntaxError: Unexpected token N in JSON at position 0
2020-07-16T14:50:35.8648780Z     at JSON.parse (<anonymous>)
2020-07-16T14:50:35.8650070Z     at runFastlaneAsync (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload/utils.ts:76:20)
2020-07-16T14:50:35.8650390Z     at processTicksAndRejections (internal/process/task_queues.js:97:5)
2020-07-16T14:50:35.8651280Z     at IOSUploader._uploadToTheStore (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload/IOSUploader.ts:225:5)
2020-07-16T14:50:35.8652530Z     at IOSUploader.upload (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload/BaseUploader.ts:33:5)
2020-07-16T14:50:35.8653440Z     at /Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/commands/upload.ts:119:9
2020-07-16T14:50:35.8654480Z     at Command.<anonymous> (/Users/runner/hostedtoolcache/expo-cli/3.22.3/x64/node_modules/expo-cli/src/exp.ts:80:7)
2020-07-16T14:50:35.8754750Z make: *** [upload-ios] Error 1
2020-07-16T14:50:35.8764110Z ##[error]Process completed with exit code 2.
```

## Environment

TBC - will come back with this 

## Reproducible Demo

TBC - will come back with this

### PR DESCRIPTION
**Pull request by mymattcarroll**
### Why

- closes #2368

### Test plan

follow the steps in #2368, this should fix the issue.

If I could get some help running this locally, that would be greatly appreciated.

### CODE REVIEW

**Comment by mymattcarroll** [comment]
@dsokal Any chance this could get a look? We would really like to get uploads consistently working again. Been a bit of a pain getting false negatives.

**Comment by brentvatne** [comment]
can you explain the logic behind these code changes? it’s hard to judge code without tests when there isn’t a strong justification for the precise code being changed

**Comment by mymattcarroll** [comment]
Of course 👍 

Using `JSON.parse()` to determine if a FastLane command has failed or succeeded is always going to be brittle logic. The FastLane CLI could release a change that outputs a non-json string. This change could even come in a patch version as I highly doubt the maintainers of the CLI consider changes to the output logs to be breaking changes.

The changes proposed are to use the exit code that is returned by the FastLane command to check if the command failed or not. This is much more robust as the exit code changing would (in my opinion) be a breaking change to the CLI. This would be much easier to maintain on Expo's side as I am sure your team would be reluctant to change major versions of the FastLane CLI without considering the breaking changes first.

**Comment by brentvatne** [comment]
@mymattcarroll - ok that makes sense. i'm going to wait for @dsokal to come back from vacation to review this because he is the owner of this code

**Comment by dsokal** [comment]
@mymattcarroll sorry it's taking so long but I just came back from vacation.

> Using JSON.parse() to determine if a FastLane command has failed or succeeded is always going to be brittle logic. The FastLane CLI could release a change that outputs a non-json string. This change could even come in a patch version as I highly doubt the maintainers of the CLI consider changes to the output logs to be breaking changes.

Actually, there is no way that "fastlane" commands could release a change that outputs a non-json string without our knowledge. The reason is simple - we're not using fastlane directly. We have our tiny wrapper called `@expo/traveling-fastlane-{darwin,linux}` which has fastlane and the appropriate ruby version bundled inside. The traveling-fastlane commands (scripts) always return JSON objects. An exception from this rule is when there is some bug in the wrapper lib. Under the hood, traveling-fastlane is executing actual fastlane commands, and those commands can potentially print something to stderr but this is not expected behavior and often means that we haven't handled some case. The root cause of this issue seems to be related to the fact that we're relying on the JSON printed to stderr. If there is some case which we haven't handled - some fastlane log can find its way to stderr and we're unable to parse it in expo-cli hence issues similar to this one.

Unfortunately, from the PR description, I can't deduce what bug you're trying to fix here. Could you please elaborate? I'd be grateful if you could take a look at the abovementioned wrapper's code - https://github.com/expo/expo-cli/tree/master/packages/traveling-fastlane and debug the problem there without touching the expo-cli code. I believe, whatever the bug is, it should be possible to fix it there.

To sum up, I'd prefer not to merge this PR in the current form.

**Comment by mhoran** [comment]
At least in my case, and in the case of #2368, the following message is being printed to stdout, causing the JSON.parse to fail:
```
`skip_waiting_for_build_processing` used and no `changelog` supplied - skipping waiting for build processing
```
This behavior seems to have changed in a "recent" version of fastlane, hence comments that "reverting back to expo-cli@3.17.24 fixes the issue."

It doesn't look like any expo-cli code sets the `changelog`, so this makes sense. I guess the message either needs to be ignored, or a changelog needs to be set.

**Comment by tadhglewis** [comment]
any updates on this one?

**Comment by mymattcarroll** [comment]
My apologies, I have also been on vacation recently.

> Actually, there is no way that "fastlane" commands could release a change that outputs a non-json string without our knowledge.

While fastlane itself may not have changed, a non-json string has been outputted from a CLI command that is being used in this code base programmatically. Relying on a json string as output from a CLI command (whether the command is code you personally wrote or fastlane itself) is always going to be brittle logic.

I would be more than happy to look into the wrapper and work out which part is logging a non-json string, but I will not have  time in the coming weeks.

In regards to the issue that needs to be fixed: https://github.com/expo/expo-cli/pull/2530#issuecomment-696379173 is the same issue that we are having.

This issue (which seems to be happening to more and more people) also explains the issue in quite alot of detail: https://github.com/expo/expo-cli/issues/2368

**Comment by EvanBacon** [comment]
`expo upload:ios` has been deprecated in favor of `eas submit -p ios`. We've also removed fastlane from expo-cli due to the unstable nature of shipping ruby packages in a node module.
